weave/trace/weave_client.py

```diff
@@ -2111,4 +2196,71 @@ def elide_display_name(name: str) -> str:
     return name
 
 
-__docspec__ = [WeaveClient, Call, CallsIter]
+class ObjectVersionCollection:
+    """A collection of object versions for a single object.
+
+    This class provides access to all versions of a specific object.
+    By default, versions are sorted by version_index in ascending order
+    (oldest to newest), unless explicitly sorted differently when created.
+    """
+
+    def __init__(self, object_id: str, versions: list[ObjSchema], client: WeaveClient):
+        """Initialize a collection of object versions.
+
+        Args:
+            object_id: The ID of the object
+            versions: List of object versions (obj schemas)
+            client: WeaveClient instance for loading objects
+        """
+        self.object_id = object_id
+        self._raw_versions = sorted(versions, key=lambda v: v.version_index)
+        self._client = client
+        self._weave_objects: dict[int, Any] = {}  # {version: object}
+
+    def __iter__(self) -> Iterator[Any]:
+        for i in range(len(self._raw_versions)):
+            yield self._get_object_at_index(i)
+
+    def __len__(self) -> int:
+        return len(self._raw_versions)
+
+    def __getitem__(self, index: int) -> Any:
+        if index < 0:
+            # Technically it can be, but it will be confusing for most users.
+            raise IndexError(
+                "Negative indexing is not supported for ObjectVersionCollection"
+            )
+        elif index >= len(self._raw_versions):
+            raise IndexError(
+                f"Index {index} out of range for collection with {len(self._raw_versions)} versions"
+            )
+        return self._get_object_at_index(index)
+
+    def __repr__(self) -> str:
+        base_class = self.base_object_class or "Unknown"
+        return f"ObjectVersionCollection(object_id='{self.object_id}', base_class='{base_class}', versions={len(self._raw_versions)})"
+
+    def _get_object_at_index(self, index: int) -> Any:
+        if index not in self._weave_objects:
+            version = self._raw_versions[index]
+            entity, project = version.project_id.split("/", 1)
+
+            ref = ObjectRef(
+                entity=entity,
+                project=project,
+                name=version.object_id,
+                _digest=version.digest,
+            )
+            self._weave_objects[index] = self._client.get(ref)
```

Comment:
I think we generally do want the "objectified" version, e.g. a Dataset if that's what it is. However, there is some weirdness because some of the useful fields, like `wb_user_id`, `created_at`, and `version_index` are at the `ObjSchema` level. Maybe the best way to solve this would be to expose those as fields on the Dataset? But otherwise I think we would want `ObjectVersionCollection` to give you a way to get the `ObjSchema` out.

Reply:
I had these included on the ObjectVersionCollection in one iteration, but removed it.  Do you think it's worth adding back to the collection level?

Reply:
Yes, I think there's a better longer timeframe solution, but maybe useful to have something like `get_schema(index)` on the `ObjectVersionCollection`?